### PR TITLE
bpo-39259: nntplib.NNTP/NNTP_SSL now reject timeout = 0

### PR DESCRIPTION
**Pull request by corona10**
https://bugs.python.org/issue39259

### CODE REVIEW

File: Lib/nntplib.py [hunk 1]
```diff
@@ -1057,6 +1057,11 @@ def __init__(self, host, port=NNTP_PORT, user=None, password=None,
             self.sock.close()
             raise
 
+    def _create_socket(self, timeout):
+        if timeout is not None and not timeout:
+            raise ValueError('Non-blocking socket (timeout=0) is not supported')
+        return socket.create_connection((self.host, self.port), timeout)
```
Comment (vstinner) [inline Lib/nntplib.py:1062]: You should move `sys.audit("nntplib.connect", self, host, port)` here :-)

File: Lib/nntplib.py [hunk 2]
```diff
@@ -1074,27 +1079,14 @@ def __init__(self, host, port=NNTP_SSL_PORT,
...
-            try:
-                _NNTPBase._close(self)
-            finally:
-                self.sock.close()
+            self.ssl_context = ssl_context
+            NNTP.__init__(self, host, port, user, password, readermode,
+                          usenetrc, timeout)
+
+        def _create_socket(self, timeout):
+            self.sock = NNTP._create_socket(self, timeout)
```
Comment (vstinner) [inline Lib/nntplib.py]:
```suggestion
            self.sock = super()._create_socket(timeout)
```

File: Lib/nntplib.py [hunk 3]
```diff
@@ -1074,27 +1079,14 @@ def __init__(self, host, port=NNTP_SSL_PORT,
...
-                self.sock.close()
-                raise
-
-        def _close(self):
-            try:
-                _NNTPBase._close(self)
-            finally:
-                self.sock.close()
+            self.ssl_context = ssl_context
+            NNTP.__init__(self, host, port, user, password, readermode,
```
Comment (vstinner) [inline Lib/nntplib.py]:
```suggestion
            super().__init__(host, port, user, password, readermode,
```

File: Lib/nntplib.py [hunk 4]
```diff
@@ -1074,27 +1079,14 @@ def __init__(self, host, port=NNTP_SSL_PORT,
...
+                          usenetrc, timeout)
+
+        def _create_socket(self, timeout):
+            self.sock = NNTP._create_socket(self, timeout)
+            self.sock = _encrypt_on(self.sock, self.ssl_context, self.host)
```
Comment (vstinner) [inline Lib/nntplib.py]: `try: ... except: self.sock.close() raise` is needed here.

**Review by vstinner — CHANGES_REQUESTED** [review]
This refactoring is interesting. I merged your other change to reject timeout=0, you should now rebase this PR.

Currently, this PR is a refactoring **and** rejects timeout=0. I prefer to have one commit per change.

**Review by corona10 — COMMENTED** [review]
I have made the requested changes; please review again

Thanks for the review 👍 
- I 've separated commits :)
- Also, applied feedback from code reviews, Thank you very much!

**Comment by vstinner** [comment]
You should now rebase your PR on master, since I merged your other PR.

**Comment by corona10** [comment]
@vstinner 
Hmm.. I already rebase this PR.

```git

commit 38bdc38ff90fb058ae80a78ca0ecf0096f4e616c (HEAD -> bpo-39259-nntplib, corona10/bpo-39259-nntplib)
Author: Dong-hee Na <donghee.na92@gmail.com>
Date:   Sat Jan 11 00:37:14 2020 +0900

    bpo-39259: Update

commit 0aa665bb0980f81d13480c45ac534235c4d5315a
Author: Dong-hee Na <donghee.na92@gmail.com>
Date:   Sat Jan 11 00:33:02 2020 +0900

    bbpo-39259: nntplib.NNTP/NNTP_SSL now reject timeout = 0

commit 02b3045015efb1bbb9dbb314fc2217378d4c6b32
Author: Dong-hee Na <donghee.na92@gmail.com>
Date:   Fri Jan 10 23:27:45 2020 +0900

    bpo-39259: nntplib.NNTP/NNTP_SSL refactoring

commit c39b52f1527868c7ada9385669c38edf98858921 (origin/master)
Author: Dong-hee Na <donghee.na92@gmail.com>
Date:   Fri Jan 10 23:34:05 2020 +0900

    bpo-39259: poplib now rejects timeout = 0 (GH-17912)
```

**Review by vstinner — CHANGES_REQUESTED** [review]
Ooooh, now I get it :-) I merged a PR to modify poplib, but this is a PR for nntplib :-D

Please split your PR to extract the _create_socket() refactoring. I would like to first have a PR for the refactoring, and *then* have a PR to reject timeout=0. Having a single PR for two changes makes review harder. In Python, we always squash multiple commits into a single commit on merge.

**Comment by corona10** [comment]
@vstinner 
Thanks for the guide.
I 've opened a new PR https://github.com/python/cpython/pull/17939 for the refactoring work.

**Comment by corona10** [comment]
So this PR should be reviewed after #17939 is merged.

**Comment by vstinner** [comment]
I merged your PR #17939. You can now rebase this one on top of it (master branch).

**Comment by corona10** [comment]
> I merged your PR #17939. You can now rebase this one on top of it (master branch).

Thanks! I 've rebased the PR :)

**Comment by vstinner** [comment]
Thanks to the refactoring, it was way easier to review this PR. It's now merged, thanks!

**Pull request merged**